### PR TITLE
[multistage] bridge v2 query engine for leaf stage v1 multi-value column

### PR DESCRIPTION
**Pull request by xiangfu0**
Introduces a new function `arrayToMultiValue` to  bridge v2 query engine for leaf stage v1 group by multi-value column.

For query: `SELECT count(*), RandomAirports FROM airlineStats GROUP BY RandomAirports`, the plan is 
```
Execution Plan
LogicalProject(EXPR$0=[$1], RandomAirports=[$0])
  LogicalAggregate(group=[{0}], agg#0=[COUNT($1)])
    PinotLogicalExchange(distribution=[hash[0]])
      LogicalAggregate(group=[{71}], agg#0=[COUNT()])
        LogicalTableScan(table=[[airlineStats]])
```
The stacktrace is:
```
2023/07/15 20:12:31.275 ERROR [MailboxSendOperator] [query_intermediate_worker_on_52712_port-10-thread-2] Exception while transferring data on opChain: 16918855000000008_1_2
java.lang.IllegalStateException: Incompatible selection result data schema:  Expected: [RandomAirports(STRING_ARRAY),$f1(LONG)]. Actual: [RandomAirports(STRING),count(*)(LONG)]
	at com.google.common.base.Preconditions.checkState(Preconditions.java:512) ~[guava-32.0.1-jre.jar:?]
	at org.apache.pinot.query.runtime.operator.LeafStageTransferableBlockOperator.composeGroupByTransferableBlock(LeafStageTransferableBlockOperator.java:198) ~[classes/:?]
	at org.apache.pinot.query.runtime.operator.LeafStageTransferableBlockOperator.composeTransferableBlock(LeafStageTransferableBlockOperator.java:158) ~[classes/:?]
	at org.apache.pinot.query.runtime.operator.LeafStageTransferableBlockOperator.getNextBlock(LeafStageTransferableBlockOperator.java:114) ~[classes/:?]
	at org.apache.pinot.query.runtime.operator.MultiStageOperator.nextBlock(MultiStageOperator.java:57) ~[classes/:?]
	at org.apache.pinot.query.runtime.operator.MailboxSendOperator.getNextBlock(MailboxSendOperator.java:124) [classes/:?]
	at org.apache.pinot.query.runtime.operator.MultiStageOperator.nextBlock(MultiStageOperator.java:57) [classes/:?]
	at org.apache.pinot.query.runtime.operator.MultiStageOperator.nextBlock(MultiStageOperator.java:33) [classes/:?]
	at org.apache.pinot.query.runtime.executor.OpChainSchedulerService$1.runJob(OpChainSchedulerService.java:94) [classes/:?]
	at org.apache.pinot.core.util.trace.TraceRunnable.run(TraceRunnable.java:40) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

After the bridge, query is:

`SELECT count(*), arrayToMultiValue(RandomAirports) FROM airlineStats GROUP BY arrayToMultiValue(RandomAirports)`

The plan is 
```
Execution Plan
LogicalProject(EXPR$0=[$1], EXPR$1=[$0])
  LogicalAggregate(group=[{0}], agg#0=[COUNT($1)])
    PinotLogicalExchange(distribution=[hash[0]])
      LogicalAggregate(group=[{0}], agg#0=[COUNT()])
        LogicalProject($f0=[ARRAYTOMULTIVALUE($71)])
          LogicalTableScan(table=[[airlineStats]])
```
image

Also it supports predicate for multi-value in v2:
```
select RandomAirports from airlineStats WHERE arrayToMultiValue(RandomAirports) ='SEA' limit 10
```
image

### CODE REVIEW

**Comment by kishoreg** [comment]
it feels like the other way around right? multivalue_to_array?

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java [hunk 1]
```diff
@@ -216,5 +216,10 @@ public static Object clpDecode(String logtypeFieldName, String dictVarsFieldName
         String defaultValue) {
       throw new UnsupportedOperationException("Placeholder scalar function, should not reach here");
     }
+
+    @ScalarFunction(names = {"arrayToMultiValue", "array_to_multi_value"}, isPlaceholder = true)
+    public static String arrayToMultiValue(Object multiValue) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:221]: (minor) Personally prefer `arrayToMV` to be consistent with other MV functions such as `sumMV`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.pinot.core.operator.blocks.ValueBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.segment.spi.index.reader.Dictionary;
+import org.roaringbitmap.RoaringBitmap;
+
+
+/**
+ * <code>ArrayToMultiValueTransformFunction</code> is the bridge for leaf stage Group By Multi-Value column.
+ */
+public class ArrayToMultiValueTransformFunction implements TransformFunction {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java:36]: Is this used in v1 or v2? Is it possible to just remove this function on the leaf?
With the current approach, filter will be very inefficient even if it works

Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java:36]: In v1 leaf, this is just the transparent layer from underlying function

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java:222]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java]: Understood, but with transform function no index can be used in filter

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java [hunk 3]
```diff
@@ -485,6 +486,25 @@ public void testLiteralOnlyFunc()
     assertEquals(results.get(10).asText(), "hello!");
   }
 
+  @Test
+  public void testMultiValueColumnGroupBy()
+      throws Exception {
+    String pinotQuery = "SELECT count(*), arrayToMV(RandomAirports) FROM mytable "
```
Comment (walterddr) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:492]: BaseIntegrationTestSets also have (1) hard-coded MV tests queries; and (2) withMultiValue flag in generated queries
should we turn them on?

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 4]
```diff
@@ -90,6 +90,11 @@ public enum TransformFunctionType {
   // date type conversion functions
   CAST("cast"),
 
+  // object type
+  ARRAY_TO_MV("arrayToMV",
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:94]: IMO this is confusing name. the data type is already MV running an ARRAY_TO_MV is a bit weird. 
should we named it `USE_AS_MV`? and we can say that MV columns are by default `USE_AS_ARRAY`

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java [hunk 5]
```diff
@@ -485,6 +486,25 @@ public void testLiteralOnlyFunc()
...
+        + "GROUP BY arrayToMV(RandomAirports)";
+    JsonNode jsonNode = postQuery(pinotQuery);
+    Assert.assertEquals(jsonNode.get("resultTable").get("rows").size(), 154);
+  }
+
+  @Test
+  public void testMultiValueColumnGroupByOrderBy()
+      throws Exception {
+    String pinotQuery = "SELECT count(*), arrayToMV(RandomAirports) FROM mytable "
```
Comment (walterddr) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:501]: will it work if I run 
```
SELECT count(*), arrayToMV(RandomAirports) 
FROM mytable 
WHERE Dest IN (SELECT Dest FROM myTable GROUP BY Dest HAVING count(*) > 10)
```
(later when we implemented the scalar function wrapper)

**Review by walterddr — COMMENTED** [review]
the grand question is whether we continue to support MV as a data type or collapse into ARRAY. 
if the answer to the above question is NO, this PR is not needed. 
if the answer is YES. there are several concerns. i thought about it a bit last night. the problem is 2 fold.
1. how do we support it conforming with ANSI syntax
2. how do we make sure v1 perform the same as before. 

for Q1: the answer is probably MULTI_SET 
- it is a SET with no ordering but can contain duplicates, suitable for our use case for GROUP-BY and FILTER, which they are considered as expanded/unnested during eval. 
- but not every use cases is considered as MULTI_SET, for example selection will consider MV as an ARRAY
therefore, 
1. an explicit `USE_AS_MULTISET(mv)` is desired to bridge the syntatic gap
2. explicit `USE_AS_ARRAY(mv)` is the default (considering select * there's no reason to ask user explicitly put this in

for Q2: the problem is these USE_AS_*** methods are considered transform, which shouldn't be the case there's a simply way to solve this --> in CalciteRexExpressionUtils, we can simply ignored the functionCall and directly return the operand as a reference in V1 
b/c V1 is SqlNode and bares no type info, this means directly putting the operand input reference without the USE_AS_*** type conversion will work naturally with the V1 context. 

so in short my proposal is
1. have `USE_AS_ARRAY` and `USE_AS_MULTISET` as scalarFunction wrappers (unimplemented) to bridge the gap of the syntactic problem on calcite
2. in PhysicalPlanner explicitly exclude these syntatitic functions and directly drop to the operand; 
3. in the mid term we will have MV, MULTI_SET, ARRAY as 3 "co-existing" types, but in V1 there's only MV; in V2 MV is considered by default as ARRAY, unless `USE_AS_MULTISET` is used. 
4. in the long term, once we fully found all alternatives in standard SQL syntax for MV, we can safely consider MV as ARRAY. and only do MULTISET when necessary (i can only see it being useful in `MV > 5` situation, where there's no array equivalent, only multiset equivalent) 

WDYT? Please let me know

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java (hunk 5 above)
Comment (xiangfu0) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:501]: You need a `group by arrayToMV(RandomAirports)` ?

Comment (xiangfu0) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:501]: This only works at leaf stage not intermediate stage.

**Comment by xiangfu0** [comment]
> the grand question is whether we continue to support MV as a data type or collapse into ARRAY. if the answer to the above question is NO, this PR is not needed. if the answer is YES. there are several concerns. i thought about it a bit last night. the problem is 2 fold.
> 
> 1. how do we support it conforming with ANSI syntax
> 2. how do we make sure v1 perform the same as before.
> 
> for Q1: the answer is probably MULTI_SET
> 
> * it is a SET with no ordering but can contain duplicates, suitable for our use case for GROUP-BY and FILTER, which they are considered as expanded/unnested during eval.
> * but not every use cases is considered as MULTI_SET, for example selection will consider MV as an ARRAY
>   therefore,
> 
> 1. an explicit `USE_AS_MULTISET(mv)` is desired to bridge the syntatic gap
> 2. explicit `USE_AS_ARRAY(mv)` is the default (considering select * there's no reason to ask user explicitly put this in
> 
> for Q2: the problem is these USE_AS_*** methods are considered transform, which shouldn't be the case there's a simply way to solve this --> in CalciteRexExpressionUtils, we can simply ignored the functionCall and directly return the operand as a reference in V1 b/c V1 is SqlNode and bares no type info, this means directly putting the operand input reference without the USE_AS_*** type conversion will work naturally with the V1 context.
> 
> so in short my proposal is
> 
> 1. have `USE_AS_ARRAY` and `USE_AS_MULTISET` as scalarFunction wrappers (unimplemented) to bridge the gap of the syntactic problem on calcite
> 2. in PhysicalPlanner explicitly exclude these syntatitic functions and directly drop to the operand;
> 3. in the mid term we will have MV, MULTI_SET, ARRAY as 3 "co-existing" types, but in V1 there's only MV; in V2 MV is considered by default as ARRAY, unless `USE_AS_MULTISET` is used.
> 4. in the long term, once we fully found all alternatives in standard SQL syntax for MV, we can safely consider MV as ARRAY. and only do MULTISET when necessary (i can only see it being useful in `MV > 5` situation, where there's no array equivalent, only multiset equivalent)
> 
> WDYT? Please let me know

MV column in V2 is modeled as ARRAY by default.
So in terms of supporting v1 format, we need to:
1. use `ARRAY_TO_MV` or `USE_AS_MV` as the bridge to ensure Calcite understand the type consistency/conversion.
2. (TODO) Implement MV GROUP BY in v2 intermediate stage to complete the story.
3. (TODO) Implement ARRAY GROUP BY later on

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:492]: v1/v2 behavior are different, we can turn on them.

Comment (xiangfu0) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:492]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ArrayToMultiValueTransformFunction.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:94]: `USE_AS_MV` might be confusing, since we repurposed `MV` as `ARRAY` in v2, `ARRAY_TO_MV` might be more explicit.

**Comment by xiangfu0** [comment]
Test failures requiring fix by https://github.com/apache/pinot/pull/11151

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java (hunk 5 above)
Comment (walterddr) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:501]: Ok sounds good

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 6]
```diff
@@ -90,6 +90,11 @@ public enum TransformFunctionType {
...
+      ReturnTypes.cascade(opBinding -> positionalComponentReturnType(opBinding, 0), SqlTypeTransforms.FORCE_NULLABLE),
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:95]: let's first create a component return type registry on @ScalarFunction so we dont have to modify the transform function side.

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:95]: actually ignored previous comment, did a bit of research and it seems like registering TransformFunctionType without an actual impl is better than having to parse scalar function annotation, which doesn't really allow anything other than primitives

**Comment by xiangfu0** [comment]
rebase to master

File: pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java [hunk 7]
```diff
@@ -199,6 +203,12 @@ private static Expression compileFunctionExpression(RexExpression.FunctionCall r
         return compileOrExpression(rexCall, pinotQuery);
       case OTHER_FUNCTION:
         functionName = rexCall.getFunctionName();
+        // Special handle for leaf stage multi-value columns, as the default behavior for filter and group by is not
+        // sql standard, so need to use `array_to_mv` to convert the array to v1 multi-value column for behavior
+        // consistency meanwhile not violating the sql standard.
+        if (ARRAY_TO_MV_ALIAS.contains(functionName)) {
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java:209]: Is function name always upper case?

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java [hunk 8]
```diff
@@ -50,6 +52,8 @@
 public class CalciteRexExpressionParser {
   private static final Logger LOGGER = LoggerFactory.getLogger(CalciteRexExpressionParser.class);
   private static final Map<String, String> CANONICAL_NAME_TO_SPECIAL_KEY_MAP;
+  private static final Set<String> ARRAY_TO_MV_ALIAS =
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java:55]: Do we need this many variance? Once we have unnest this should be removed right?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java (hunk 7 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java:209]: Let's canonicalized and just do `ARRAY_TO_MV`

**Review by walterddr — APPROVED** [review]
Lgtm

File: pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java (hunk 8 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/parser/CalciteRexExpressionParser.java]: done

**Comment by xiangfu0** [comment]
Part of this https://github.com/apache/pinot/issues/10658
